### PR TITLE
Ability to validate not only the file and all files and files in a folder [Script Validator]

### PR DESCRIPTION
**Pull request by Alexandr57**
Ability to validate not only the file and all files and files in a folder

### CODE REVIEW

**Comment by Alexandr57** [comment]
image

File: Utils/ScriptValidator/Unit1.pas [hunk 1]
```diff
@@ -30,21 +34,37 @@   TForm1 = class(TForm)
 
 var
   Form1: TForm1;
-
-
+  IsValidatePath : Boolean;
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Does not need to be global vars

File: Utils/ScriptValidator/Unit1.pas [hunk 2]
```diff
@@ -30,21 +34,37 @@   TForm1 = class(TForm)
...
-
+  IsValidatePath : Boolean;
+  sListFileInFolder : TStringList;
 implementation
 uses
   KM_Maps;
 
 {$R *.dfm}
 
+procedure TForm1.FindFiles(Path: String; out aList: TStringList);
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Why do you add "out" modifier? It should not be here

Also please mind the code standard - arguments start with "a" = aPath

File: Utils/ScriptValidator/Unit1.pas [hunk 3]
```diff
@@ -30,21 +34,37 @@   TForm1 = class(TForm)
...
+var SearchRec:TSearchRec;
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Please mind the code formatting. This should be:

var
  SearchRec: TSearchRec;

File: Utils/ScriptValidator/Unit1.pas [hunk 4]
```diff
@@ -30,21 +34,37 @@   TForm1 = class(TForm)
...
+begin
+  FindFirst(Path+PathDelim+'*', faAnyFile, SearchRec);
+  repeat
+    if (SearchRec.Name <> '.') and (SearchRec.Name <> '..') then
+      if (SearchRec.Attr and faDirectory = faDirectory) then
+        FindFiles(Path + PathDelim + SearchRec.Name, aList)
+      else
+        if SameText(RightStr(SearchRec.Name, Length(EXT_FILE_SCRIPT)+1), ('.'+EXT_FILE_SCRIPT)) then
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Looks quite complicated.
Should be SameText(ExtractFileExt(SearchRec.Name), EXT_FILE_SCRIPT)

File: Utils/ScriptValidator/Unit1.pas [hunk 5]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+  sListFileInFolder.Free;
   DragAcceptFiles(Handle, False);
 end;
 
-procedure TForm1.btnBrowseClick(Sender: TObject);
+procedure TForm1.btnBrowseFileClick(Sender: TObject);
 begin
   if OpenDialog1.Execute then
-    Edit1.Text := OpenDialog1.FileName;
+  Edit1.Text := OpenDialog1.FileName;
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Mind code formatting please

File: Utils/ScriptValidator/Unit1.pas [hunk 6]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
 end;
 
 
+procedure TForm1.btnBrowsePathClick(Sender: TObject);
+var DirValidate : String;
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Mind code formatting please

File: Utils/ScriptValidator/Unit1.pas [hunk 7]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+begin
+  If (GetFileAttributes(PWideCHar(TEdit(Sender).Text)) and FILE_ATTRIBUTE_DIRECTORY) = FILE_ATTRIBUTE_DIRECTORY then
+  begin
+    IsValidatePath := true;
+    Memo1.Text := 'Selected folders!';
+  end
+  else
+  begin
+    IsValidatePath := false;
+    Memo1.Text := 'Select Files!';
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Why is the text different, there you reference selectED folders, and here select files?

File: Utils/ScriptValidator/Unit1.pas [hunk 8]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+
 procedure TForm1.btnValidateClick(Sender: TObject);
+var I : Integer;
 begin
-  Validate(Edit1.Text, True);
+  btnBrowsePath.Enabled := false;
+  btnBrowseFile.Enabled := false;
+  btnValidate.Enabled := false;
+  btnValidateAll.Enabled := false;
+  Edit1.Enabled := false;
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Would be better to move these 5 lines into a separate procedure.

File: Utils/ScriptValidator/Unit1.pas [hunk 9]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+
+  Memo1.Lines.Clear;
+  if IsValidatePath then
+  begin
+    if SameText(RightStr(Edit1.Text, 1), PathDelim) then Edit1.Text := LeftStr(Edit1.Text, (Length(Edit1.Text) -1));
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Should be ExcludeTrailingPathDelimiter(Edit1.Text)

File: Utils/ScriptValidator/Unit1.pas [hunk 10]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+      begin
+        Memo1.Lines.Append('Files in the folder: '+IntToStr(sListFileInFolder.Count));
+        for I := 0 to sListFileInFolder.Count-1 do
+          Validate(sListFileInFolder.Strings[I], True);
+        Memo1.Lines.Append('Checked ' + IntToStr(sListFileInFolder.Count) + ' in ..\..\');
+      end;
+
+    end;
+
+  end else
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: No need for 2 extra empty lines above

File: Utils/ScriptValidator/Unit1.pas [hunk 11]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+      FindFiles(Edit1.Text, sListFileInFolder);
+      Memo1.Lines.Append('Check '+Edit1.Text);
+      if sListFileInFolder.Count = 0 then
+        Memo1.Lines.Append('No files in a directory ' + Edit1.Text)
+      else
+      begin
+        Memo1.Lines.Append('Files in the folder: '+IntToStr(sListFileInFolder.Count));
+        for I := 0 to sListFileInFolder.Count-1 do
+          Validate(sListFileInFolder.Strings[I], True);
+        Memo1.Lines.Append('Checked ' + IntToStr(sListFileInFolder.Count) + ' in ..\..\');
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Why do you add "' in ..\..\'" here?

File: Utils/ScriptValidator/Unit1.pas [hunk 12]
```diff
@@ -53,45 +73,128 @@ procedure TForm1.FormCreate(Sender: TObject);
...
+  btnBrowsePath.Enabled := true;
+  btnBrowseFile.Enabled := true;
+  btnValidate.Enabled := true;
+  btnValidateAll.Enabled := true;
+  Edit1.Enabled := true;
 end;
 
 
-procedure TForm1.Button1Click(Sender: TObject);
+procedure TForm1.btnValidateAllClick(Sender: TObject);
```
Comment (Kromster80) [inline Utils/ScriptValidator/Unit1.pas]: Below function looks like to have a lot of duplicate code with above. Are you sure they can not be merged?

File: src/KM_Defaults.pas [hunk 13]
```diff
@@ -220,6 +220,7 @@ interface
   DOWNLOADED_LOBBY_SAVE = 'downloaded';
 
   MP_MINIMAP_SAVE_EXT = 'smm';
+  EXT_FILE_SCRIPT = 'script';
```
Comment (Kromster80) [inline src/KM_Defaults.pas]: Extensions should include the dot - ".script"

**Review by Kromster80 — CHANGES_REQUESTED** [review]

**Comment by Alexandr57** [comment]
Этот PR возможно нуждается в исправлении. Сейчас протестировать нет возможности.

This PR may need correction. Now there is no way to test it.

**Comment by Alexandr57** [comment]
Короче я проверил. Все работает. PR можете принимать или закрывать.
Делать возможность проверки каталога включая подкаталогов в консольном режиме я не буду. Пускай @thibmo этим занимается.
Как по мне проще было бы прикрутить консольный режим в графическом виде чем в консольном используя ParamStr() при создании формы.

In short, I checked. Everything is working. PR can accept or close.
Make the ability to check the directory including subdirectories in the console mode I will not. Let @thibmo do it.
As for me it would be easier to tighten the console mode in a graphical form than in the console using ParamStr () when creating the form.

**Comment by LauraRozier** [comment]
We can always change the implementation of the console/GUI mode switch. :)
Issue with GUI mode is that it calls the Windows API in a different way, meaning it will always spawn a GUI, now imagine a map scripter that works with Linux CLI and only uses the GUI for gaming..
Making a console in the GUI is a no-go in this case.
Hence the console-first approach. :)

As for the edits, sure can implement those if needed. :)

Quick explanation:
Console mode detection "magic" is done with this code:
```Pascal
function IsConsoleMode: Boolean;
var
  SI: TStartupInfo;
begin
  SI.cb := SizeOf(StartUpInfo);
  GetStartupInfo(SI);
  Result := (SI.dwFlags and STARTF_USESHOWWINDOW) = 0;
end;
```
So it SHOULD still start with GUI by default when you start the app whilst dragging a file to the exe.
(*
As in, start the PE with a file as first argument after the exe itself, caused by a drag&drop action, like so:
drag&drop PE start example
*)
The '-g' / '-graphic' flags are only there to be able to force GUI mode when started from CMD/PowerShell.

**Comment by Kromster80** [comment]
Abandoned

**Pull request closed without merging**